pkg/registry/apis/provisioning/jobs/queue.go

```diff
@@ -0,0 +1,180 @@
+package jobs
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/google/uuid"
+
+	provisioning "github.com/grafana/grafana/pkg/apis/provisioning/v0alpha1"
+)
+
+func NewJobQueue(capacity int) JobQueue {
+	return &jobStore{
+		rv:       1,
+		capacity: capacity,
+		jobs:     []provisioning.Job{},
+	}
+}
+
+type jobStore struct {
+	capacity int
+	workers  []Worker
+
+	// All jobs
+	jobs []provisioning.Job
+	rv   int64 // updates whenever changed
+
+	mutex sync.RWMutex
+}
+
+// Register a worker (inline for now)
+func (s *jobStore) Register(worker Worker) {
+	s.mutex.Lock()
+	s.workers = append(s.workers, worker)
+	s.mutex.Unlock()
+}
+
+// Add implements JobQueue.
+func (s *jobStore) Add(ctx context.Context, job provisioning.Job) (string, error) {
+	if job.Namespace == "" {
+		return "", fmt.Errorf("expecting namespace")
+	}
+	_, ok := job.Labels["repository"]
+	if !ok {
+		return "", fmt.Errorf("expecting repository name label")
+	}
+	_, ok = job.Labels["repository.type"] // TODO: ideally just based on the action type
+	if !ok {
+		return "", fmt.Errorf("expecting repository name label")
+	}
+	if job.Spec.Action == "" {
+		return "", fmt.Errorf("missing spec.action")
+	}
+
+	// Only for add
+	if job.Status.State != "" {
+		return "", fmt.Errorf("must add jobs with empty state")
+	}
+	job.Status.State = "pending" // start pending
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.rv++
+	job.ResourceVersion = strconv.FormatInt(s.rv, 10)
+	job.Labels["action"] = string(job.Spec.Action) // make it searchable
+	job.Labels["state"] = job.Status.State
+	job.Status.State = "pending" // it is queued
+	job.Name = "j" + uuid.NewString()
+	job.CreationTimestamp = metav1.NewTime(time.Now())
+
+	jobs := make([]provisioning.Job, 0, len(s.jobs)+2)
+	jobs = append(jobs, job)
+	for i, j := range s.jobs {
+		if i >= s.capacity {
+			break
+		}
+		jobs = append(jobs, j)
+	}
+
+	go func() {
```

Comment:
simple approach for now.... when we add to the queue, we kick off a background process that will process things